File: src/algoritmos.cpp
#include "solucionador_labirinto/algoritmos.hpp"
#include <queue>
#include <map>

namespace solucionador {

MapaProcessado processar_mapa(const std::vector<char>& lista_achatada,
                              int largura, int altura) {
    MapaProcessado resultado;
    resultado.inicio_encontrado = false;
    resultado.alvo_encontrado = false;
    resultado.mapa.resize(altura);
    
    for (int i = 0; i < altura; i++) {
        // pega o pedaço da lista que corresponde a uma linha
        int inicio_linha = i * largura;
        int fim_linha = (i + 1) * largura;
        
        std::vector<char> linha;
        for (int j = inicio_linha; j < fim_linha; j++) {
            linha.push_back(lista_achatada[j]);
        }
        resultado.mapa[i] = linha;
        
        // procura robô e alvo nesta linha
        for (int j = 0; j < static_cast<int>(linha.size()); j++) {
            if (linha[j] == 'r') {
                resultado.inicio = {i, j};
                resultado.inicio_encontrado = true;
            }
            if (linha[j] == 't') {
                resultado.alvo = {i, j};
                resultado.alvo_encontrado = true;
            }
        }
    }
    
    return resultado;
}

std::vector<std::string> bfs_encontrar_caminho(
    const std::vector<std::vector<char>>& mapa,
    const Posicao& inicio,
    const Posicao& alvo) {
    
    int linhas = mapa.size();
    int colunas = mapa[0].size();
    
    // fila armazena: (posicao_atual, caminho_percorrido)
    std::queue<std::pair<Posicao, std::vector<std::string>>> fila;
    fila.push({inicio, {}});
    
    std::set<Posicao> visitados;
    visitados.insert(inicio);
    
    // dicionário: (mudança_linha, mudança_coluna) -> comando_texto
    std::map<std::pair<int, int>, std::string> direcoes = {
        {{-1, 0}, "up"},    // cima diminui a linha
        {{1, 0}, "down"},   // baixo aumenta a linha
        {{0, -1}, "left"},  // esquerda diminui a coluna
        {{0, 1}, "right"}   // direita aumenta a coluna
    };
    
    while (!fila.empty()) {
        auto [pos_atual, caminho] = fila.front();
        fila.pop();
        
        // chegou no alvo?
        if (pos_atual == alvo) {
            return caminho;
        }
        
        // tenta andar para os vizinhos
        for (const auto& [delta, comando] : direcoes) {
            int nova_lin = pos_atual.linha + delta.first;
            int nova_col = pos_atual.coluna + delta.second;
            Posicao nova_pos = {nova_lin, nova_col};
            
            // verifica se não saiu do mapa
            if (nova_lin >= 0 && nova_lin < linhas && 
                nova_col >= 0 && nova_col < colunas) {
                
                // verifica se não é parede ('b') e se não visitou ainda
                if (mapa[nova_lin][nova_col] != 'b' && 
                    visitados.find(nova_pos) == visitados.end()) {
                    
                    visitados.insert(nova_pos);
                    auto novo_caminho = caminho;
                    novo_caminho.push_back(comando);
                    fila.push({nova_pos, novo_caminho});
                }
            }
        }
    }
    
    return {}; // retorna lista vazia se não achar nada
}

Explorador::Explorador(int largura, int altura)
    : largura_(largura), altura_(altura) {
    
    // inicializa tudo como '?'
    mapa_conhecido_.resize(altura_, std::vector<char>(largura_, '?'));
    
    // posição inicial
    pos_atual_ = {1, 1};
    mapa_conhecido_[1][1] = 'f';
    
    visitados_.insert(pos_atual_);
}

void Explorador::atualizar_mapa(const std::string& up, const std::string& down,
                                const std::string& left, const std::string& right) {
    int lin = pos_atual_.linha;
    int col = pos_atual_.coluna;
    
    struct Vizinho {
        int linha;
        int coluna;
        std::string valor;
    };
    
    std::vector<Vizinho> vizinhos = {
        {lin - 1, col, up},
        {lin + 1, col, down},
        {lin, col - 1, left},
        {lin, col + 1, right}
    };
    
    for (const auto& viz : vizinhos) {
        // verifica limites antes de gravar no mapa
        if (viz.linha >= 0 && viz.linha < altura_ &&
            viz.coluna >= 0 && viz.coluna < largura_) {
            mapa_conhecido_[viz.linha][viz.coluna] = viz.valor[0];
        }
    }
}

std::optional<std::string> Explorador::proximo_passo() {
    int lin = pos_atual_.linha;
    int col = pos_atual_.coluna;
    
    struct Movimento {
        int delta_lin;
        int delta_col;
        std::string direcao;
    };
    
    std::vector<Movimento> movimentos = {
        {-1, 0, "up"},
        {0, 1, "right"},
        {1, 0, "down"},
        {0, -1, "left"}
    };
    
    // 1. tenta avançar
    for (const auto& mov : movimentos) {
        int viz_l = lin + mov.delta_lin;
        int viz_c = col + mov.delta_col;
        
        // verifica se o vizinho existe antes de acessar a lista
        bool dentro_do_mapa = (viz_l >= 0 && viz_l < altura_) &&
                             (viz_c >= 0 && viz_c < largura_);
        
        if (dentro_do_mapa) {
            // só entra aqui se a coordenada for válida
            bool eh_parede = (mapa_conhecido_[viz_l][viz_c] == 'b');
            Posicao viz_pos = {viz_l, viz_c};
            bool foi_visitado = (visitados_.find(viz_pos) != visitados_.end());
            
            if (!eh_parede && !foi_visitado) {
                // acha um caminho novo
                pilha_.push(pos_atual_);
                visitados_.insert(viz_pos);
                pos_atual_ = viz_pos;
                return mov.direcao;
            }
        }
    }
    
    // 2. backtracking (voltar)
    if (!pilha_.empty()) {
        Posicao anterior = pilha_.top();
        pilha_.pop();
        
        int delta_l = anterior.linha - lin;
        int delta_c = anterior.coluna - col;
        
        pos_atual_ = anterior;
        
        if (delta_l == -1) return "up";
        if (delta_l == 1) return "down";
        if (delta_c == -1) return "left";
        if (delta_c == 1) return "right";
    }
    
    return std::nullopt;
}

const std::vector<std::vector<char>>& Explorador::get_mapa_conhecido() const {
    return mapa_conhecido_;
}

const Posicao& Explorador::get_posicao_atual() const {
    return pos_atual_;
}

} // namespace solucionador

File: solucionador_labirinto/algoritmos.py
from collections import deque

def processar_mapa(lista_achatada, largura, altura):
    """
    Transforma a lista de letras em uma matriz e acha onde está o robô ('r') e o alvo ('t').
    """
    mapa = []
    inicio = None
    alvo = None
    
    for i in range(altura):
        # pega o pedaço da lista que corresponde a uma linha
        inicio_linha = i * largura
        fim_linha = (i + 1) * largura
        linha = lista_achatada[inicio_linha:fim_linha]
        mapa.append(linha)
        
        # procura robô e alvo nesta linha
        if 'r' in linha:
            inicio = (i, linha.index('r')) # (linha, coluna)
        if 't' in linha:
            alvo = (i, linha.index('t'))
            
    return mapa, inicio, alvo

def bfs_encontrar_caminho(mapa, inicio, alvo):
    """
    Usa busca em largura (BFS) para achar o menor caminho.
    Retorna uma lista de comandos: ['up', 'left', 'down', ...]
    """
    linhas = len(mapa)
    colunas = len(mapa[0])
    
    # fila armazena: (posicao_atual, caminho_percorrido)
    fila = deque([ (inicio, []) ])
    visitados = set([inicio])
    
    # dicionário: (mudança_linha, mudança_coluna) -> 'comando_texto'
    direcoes = {
        (-1, 0): 'up',    # cima diminui a linha
        (1, 0): 'down',   # baixo aumenta a linha
        (0, -1): 'left',  # esquerda diminui a coluna
        (0, 1): 'right'   # direita aumenta a coluna
    }

    while fila:
        (linha_atual, col_atual), caminho = fila.popleft()

        # chegou no alvo?
        if (linha_atual, col_atual) == alvo:
            return caminho

        # tenta andar para os vizinhos
        for (delta_lin, delta_col), comando in direcoes.items():
            nova_lin, nova_col = linha_atual + delta_lin, col_atual + delta_col

            # verifica se não saiu do mapa
            if 0 <= nova_lin < linhas and 0 <= nova_col < colunas:
                # verifica se não é parede ('b') e se não visitou ainda
                if mapa[nova_lin][nova_col] != 'b' and (nova_lin, nova_col) not in visitados:
                    visitados.add((nova_lin, nova_col))
                    novo_caminho = caminho + [comando]
                    fila.append(((nova_lin, nova_col), novo_caminho))
    
    return [] # retorna lista vazia se não achar nada


class Explorador:
    def __init__(self, largura=29, altura=29):
        self.largura = largura
        self.altura = altura
        # inicializa tudo como '?'
        self.mapa_conhecido = [['?' for _ in range(largura)] for _ in range(altura)]
        
        # posição inicial
        self.pos_atual = (1, 1) 
        self.mapa_conhecido[1][1] = 'f' 
        
        self.visitados = set()
        self.visitados.add(self.pos_atual)
        
        self.pilha = [] 

    def atualizar_mapa(self, sensor_msg):
        lin, col = self.pos_atual
        
        vizinhos = [
            ('up',    lin - 1, col),
            ('down',  lin + 1, col),
            ('left',  lin,     col - 1),
            ('right', lin,     col + 1)
        ]
        
        for direcao, l, c in vizinhos:
            # verifica limites antes de gravar no mapa
            if 0 <= l < self.altura and 0 <= c < self.largura:
                valor_sensor = getattr(sensor_msg, direcao)
                self.mapa_conhecido[l][c] = valor_sensor

    def proximo_passo(self):
        lin, col = self.pos_atual
        
        movimentos = [
            ((-1, 0), 'up'),
            ((0, 1),  'right'),
            ((1, 0),  'down'),
            ((0, -1), 'left')
        ]
        
        # 1. tenta avançar
        for (d_lin, d_col), direcao_txt in movimentos:
            viz_l, viz_c = lin + d_lin, col + d_col
            
            # verifica se o vizinho existe antes de acessar a lista
            dentro_do_mapa = (0 <= viz_l < self.altura) and (0 <= viz_c < self.largura)
            
            if dentro_do_mapa:
                # só entra aqui se a coordenada for válida
                eh_parede = (self.mapa_conhecido[viz_l][viz_c] == 'b')
                foi_visitado = ((viz_l, viz_c) in self.visitados)
                
                if not eh_parede and not foi_visitado:
                    # acha um caminho novo
                    self.pilha.append(self.pos_atual)
                    self.visitados.add((viz_l, viz_c))
                    self.pos_atual = (viz_l, viz_c)
                    return direcao_txt

        # 2. backtracking (voltar)
        if self.pilha:
            anterior_l, anterior_c = self.pilha.pop()
            
            delta_l = anterior_l - lin
            delta_c = anterior_c - col
            
            self.pos_atual = (anterior_l, anterior_c)
            
            if delta_l == -1: return 'up'
            if delta_l == 1:  return 'down'
            if delta_c == -1: return 'left'
            if delta_c == 1:  return 'right'
            
        return None
